File: python/rafter_cli/commands/issues/issue_builder.py
"""Build structured GitHub issues from scan findings."""
from __future__ import annotations

import os
from dataclasses import dataclass, field

from .dedup import embed_fingerprint, fingerprint


@dataclass
class IssueDraft:
    title: str
    body: str
    labels: list[str] = field(default_factory=list)
    fingerprint: str = ""


@dataclass
class BackendVulnerability:
    rule_id: str
    level: str
    message: str
    file: str
    line: int | None = None


@dataclass
class LocalMatch:
    pattern_name: str
    severity: str
    description: str = ""
    line: int | None = None
    column: int | None = None
    redacted: str = ""


_SEVERITY_MAP = {
    "error": "critical",
    "critical": "critical",
    "warning": "high",
    "high": "high",
    "note": "medium",
    "medium": "medium",
    "low": "low",
}

_SEVERITY_EMOJI = {
    "critical": "\U0001f534",
    "high": "\U0001f7e0",
    "medium": "\U0001f7e1",
    "low": "\U0001f7e2",
}


def _severity_label(level: str) -> str:
    return _SEVERITY_MAP.get(level.lower(), "medium")


def _severity_emoji(level: str) -> str:
    return _SEVERITY_EMOJI.get(_severity_label(level), "\U0001f7e1")


def _truncate(s: str, max_len: int) -> str:
    if len(s) <= max_len:
        return s
    return s[: max_len - 3] + "..."


def build_from_backend_vulnerability(vuln: BackendVulnerability) -> IssueDraft:
    sev = _severity_label(vuln.level)
    emoji = _severity_emoji(vuln.level)
    fp = fingerprint(vuln.file, vuln.rule_id)

    title = f"{emoji} [{sev.upper()}] {vuln.rule_id}: {_truncate(vuln.message, 80)}"

    body = f"## Security Finding\n\n"
    body += f"**Rule:** `{vuln.rule_id}`\n"
    body += f"**Severity:** {sev}\n"
    body += f"**File:** `{vuln.file}`"
    if vuln.line:
        body += f" (line {vuln.line})"
    body += "\n\n"
    body += f"### Description\n\n{vuln.message}\n\n"
    body += f"### Remediation\n\nReview and fix the finding in `{vuln.file}`.\n"
    body += "\n---\n*Created by [Rafter CLI](https://rafter.so) — security for AI builders*\n"

    labels = ["security", f"severity:{sev}", f"rule:{vuln.rule_id}"]

    return IssueDraft(
        title=title,
        body=embed_fingerprint(body, fp),
        labels=labels,
        fingerprint=fp,
    )


def build_from_local_match(file: str, match: LocalMatch) -> IssueDraft:
    sev = _severity_label(match.severity)
    emoji = _severity_emoji(match.severity)
    fp = fingerprint(file, match.pattern_name)

    basename = os.path.basename(file)
    title = f"{emoji} [{sev.upper()}] Secret detected: {match.pattern_name} in {basename}"

    body = "## Secret Detection\n\n"
    body += f"**Pattern:** `{match.pattern_name}`\n"
    body += f"**Severity:** {sev}\n"
    body += f"**File:** `{file}`"
    if match.line:
        body += f" (line {match.line})"
    body += "\n"
    if match.redacted:
        body += f"**Match:** `{match.redacted}`\n"
    body += "\n"
    if match.description:
        body += f"### Description\n\n{match.description}\n\n"
    body += "### Remediation\n\n"
    body += "1. Rotate the exposed credential immediately\n"
    body += "2. Remove the secret from source code\n"
    body += "3. Use environment variables or a secrets manager instead\n"
    body += "\n---\n*Created by [Rafter CLI](https://rafter.so) — security for AI builders*\n"

    labels = ["security", "secret-detected", f"severity:{sev}"]

    return IssueDraft(
        title=title,
        body=embed_fingerprint(body, fp),
        labels=labels,
        fingerprint=fp,
    )

File: node/src/commands/issues/issue-builder.ts
/**
 * Build structured GitHub issues from scan findings.
 *
 * Handles both:
 * - Backend scan vulnerabilities (SAST/policy findings)
 * - Local scan results (secret detection)
 */
import { fingerprint, embedFingerprint } from "./dedup.js";

export interface IssueDraft {
  title: string;
  body: string;
  labels: string[];
  fingerprint: string;
}

/** Backend vulnerability from `rafter get <id> --format json` */
export interface BackendVulnerability {
  ruleId: string;
  level: string;
  message: string;
  file: string;
  line?: number;
  result?: any;
}

/** Local scan result from `rafter scan local --format json` */
export interface LocalScanResult {
  file: string;
  matches: Array<{
    pattern: { name: string; severity: string; description?: string };
    line?: number;
    column?: number;
    redacted?: string;
  }>;
}

function severityLabel(level: string): string {
  const map: Record<string, string> = {
    error: "critical",
    critical: "critical",
    warning: "high",
    high: "high",
    note: "medium",
    medium: "medium",
    low: "low",
  };
  return map[level.toLowerCase()] || "medium";
}

function severityEmoji(level: string): string {
  const sev = severityLabel(level);
  const emojis: Record<string, string> = {
    critical: "🔴",
    high: "🟠",
    medium: "🟡",
    low: "🟢",
  };
  return emojis[sev] || "🟡";
}

export function buildFromBackendVulnerability(
  vuln: BackendVulnerability
): IssueDraft {
  const sev = severityLabel(vuln.level);
  const emoji = severityEmoji(vuln.level);
  const fp = fingerprint(vuln.file, vuln.ruleId);

  const title = `${emoji} [${sev.toUpperCase()}] ${vuln.ruleId}: ${truncate(vuln.message, 80)}`;

  let body = `## Security Finding\n\n`;
  body += `**Rule:** \`${vuln.ruleId}\`\n`;
  body += `**Severity:** ${sev}\n`;
  body += `**File:** \`${vuln.file}\``;
  if (vuln.line) body += ` (line ${vuln.line})`;
  body += `\n\n`;
  body += `### Description\n\n${vuln.message}\n\n`;
  body += `### Remediation\n\nReview and fix the finding in \`${vuln.file}\`.\n`;
  body += `\n---\n*Created by [Rafter CLI](https://rafter.so) — security for AI builders*\n`;

  const labels = [
    "security",
    `severity:${sev}`,
    `rule:${vuln.ruleId}`,
  ];

  return {
    title,
    body: embedFingerprint(body, fp),
    labels,
    fingerprint: fp,
  };
}

export function buildFromLocalMatch(
  file: string,
  match: LocalScanResult["matches"][0]
): IssueDraft {
  const sev = severityLabel(match.pattern.severity);
  const emoji = severityEmoji(match.pattern.severity);
  const fp = fingerprint(file, match.pattern.name);

  const title = `${emoji} [${sev.toUpperCase()}] Secret detected: ${match.pattern.name} in ${basename(file)}`;

  let body = `## Secret Detection\n\n`;
  body += `**Pattern:** \`${match.pattern.name}\`\n`;
  body += `**Severity:** ${sev}\n`;
  body += `**File:** \`${file}\``;
  if (match.line) body += ` (line ${match.line})`;
  body += `\n`;
  if (match.redacted) {
    body += `**Match:** \`${match.redacted}\`\n`;
  }
  body += `\n`;
  if (match.pattern.description) {
    body += `### Description\n\n${match.pattern.description}\n\n`;
  }
  body += `### Remediation\n\n`;
  body += `1. Rotate the exposed credential immediately\n`;
  body += `2. Remove the secret from source code\n`;
  body += `3. Use environment variables or a secrets manager instead\n`;
  body += `\n---\n*Created by [Rafter CLI](https://rafter.so) — security for AI builders*\n`;

  const labels = [
    "security",
    "secret-detected",
    `severity:${sev}`,
  ];

  return {
    title,
    body: embedFingerprint(body, fp),
    labels,
    fingerprint: fp,
  };
}

function truncate(s: string, max: number): string {
  if (s.length <= max) return s;
  return s.slice(0, max - 3) + "...";
}

function basename(filepath: string): string {
  return filepath.split("/").pop() || filepath;
}
